File: src/lib/ids.ts
/**
 * Normalize a Notion ID — accept both dashless (from URLs) and dashed (standard UUID) formats.
 *
 * Examples:
 *   "30a61d9c1112802f95fef30d3a601ec5" → "30a61d9c-1112-802f-95fe-f30d3a601ec5"
 *   "30a61d9c-1112-802f-95fe-f30d3a601ec5" → "30a61d9c-1112-802f-95fe-f30d3a601ec5"
 */
export function normalizeId(id: string): string {
  const cleaned = id.replace(/-/g, "").toLowerCase();
  if (cleaned.length !== 32 || !/^[0-9a-f]+$/.test(cleaned)) {
    return id; // Not a valid UUID hex string — return as-is and let the API reject it
  }
  return `${cleaned.slice(0, 8)}-${cleaned.slice(8, 12)}-${cleaned.slice(12, 16)}-${cleaned.slice(16, 20)}-${cleaned.slice(20)}`;
}

File: internal/ids/ids.go
// Package ids normalizes Notion identifiers: URLs and copy-pasted IDs arrive
// dashless, the APIs want dashed UUIDs.
package ids

import "strings"

// Normalize accepts a dashless (from URLs) or dashed (standard UUID) Notion
// ID and returns the dashed form. Anything that is not 32 hex chars after
// stripping dashes is returned unchanged, letting the API reject it.
func Normalize(id string) string {
	cleaned := strings.ToLower(strings.ReplaceAll(id, "-", ""))
	if len(cleaned) != 32 || !isHex(cleaned) {
		return id
	}
	return cleaned[0:8] + "-" + cleaned[8:12] + "-" + cleaned[12:16] + "-" + cleaned[16:20] + "-" + cleaned[20:]
}

func isHex(s string) bool {
	for _, c := range s {
		if (c < '0' || c > '9') && (c < 'a' || c > 'f') {
			return false
		}
	}
	return true
}
